File: tuan02/bai43/bai43.py
dem = 0

def fib_de_quy(n: int) -> int:
    global dem
    dem += 1
    if n <= 1:
        return n
    return fib_de_quy(n - 1) + fib_de_quy(n - 2)

def fib_ghi_nho_helper(n: int, memo: list[int]) -> int:
    global dem
    dem += 1
    if n <= 1:
        return n
    if memo[n] != -1:
        return memo[n]
    memo[n] = fib_ghi_nho_helper(n - 1, memo) + fib_ghi_nho_helper(n - 2, memo)
    return memo[n]

def fib_ghi_nho(n: int) -> int:
    memo = [-1] * (n + 1)
    return fib_ghi_nho_helper(n, memo)

def fib_lap(n: int) -> int:
    if n <= 1:
        return n
    a, b = 0, 1
    for _ in range(2, n + 1):
        a, b = b, a + b
    return b

if __name__ == "__main__":
    import sys
    tokens = sys.stdin.read().split()
    if tokens:
        n = int(tokens[0])
        
        dem = 0
        val1 = fib_de_quy(n)
        call1 = dem
        
        dem = 0
        val2 = fib_ghi_nho(n)
        call2 = dem
        
        val3 = fib_lap(n)
        
        print(f"{val1} {call1} {call2}")

File: tuan02/bai43/bai43.cpp
#include <iostream>
#include <vector>

using namespace std;

long long dem = 0;

long long fib_de_quy(int n) {
    dem++;
    if (n <= 1) return n;
    return fib_de_quy(n - 1) + fib_de_quy(n - 2);
}

long long fib_ghi_nho_helper(int n, vector<long long>& memo) {
    dem++;
    if (n <= 1) return n;
    if (memo[n] != -1) return memo[n];
    memo[n] = fib_ghi_nho_helper(n - 1, memo) + fib_ghi_nho_helper(n - 2, memo);
    return memo[n];
}

long long fib_ghi_nho(int n) {
    vector<long long> memo(n + 1, -1);
    return fib_ghi_nho_helper(n, memo);
}

long long fib_lap(int n) {
    if (n <= 1) return n;
    long long a = 0, b = 1, c = 0;
    for (int i = 2; i <= n; i++) {
        c = a + b;
        a = b;
        b = c;
    }
    return c;
}

int main() {
    int n;
    if (cin >> n) {
        dem = 0;
        long long val1 = fib_de_quy(n);
        long long call1 = dem;

        dem = 0;
        long long val2 = fib_ghi_nho(n);
        long long call2 = dem;

        long long val3 = fib_lap(n);

        cout << val1 << " " << call1 << " " << call2 << endl;
    }
    return 0;
}